### DISCUSSION TITLE
Need help testing RTL languages for next major release

### DISCUSSION THREAD

**Discussion by kevinpapst** [kimai/kimai#3463] [General]
To everyone who has a right-to-left mother language, I need some help with testing the new release, which adds RTL functionality.

A lot of the work is already done (at least I think so), but there are things missing which I am not capable to solve - simply because I don't know how it should work. Most of the formatting codes are now derived from the ICU libraries and not hard-coded in Kimai anymore. Therefor it should already be much more consistent. But there are still a couple of places where I do manual formatting, e.g. for duration. 

 @yarons you are the only one I know, who definitely knows a lot about a RTL locale. Could you join the discussion and help?
For now I would need feedback on the `he` screenshots. The first one is a demo section of the new version that showcases all available format functions, which are used everywhere in Kimai to render the data (mostly dates, times, duration, money). The second one the timesheet listing.

## English (en_GB)
<img width="1082" alt="Bildschirmfoto 2022-06-26 um 14 05 56" src="https://user-images.githubusercontent.com/533162/175813310-07f1e333-b55d-4d0f-8877-e9d6302da7c9.png">
<img width="1130" alt="Bildschirmfoto 2022-06-26 um 14 10 50" src="https://user-images.githubusercontent.com/533162/175813506-2d0231d2-d72d-4e6b-8846-20541600b77c.png">

## Hebrew (he)
And now the ones where I need feedback on. Can you spot formatting (like duration) which should be improved?
The demo data, like customer names, are still written in LTR, please take that into account:
<img width="1080" alt="Bildschirmfoto 2022-06-26 um 14 06 20" src="https://user-images.githubusercontent.com/533162/175813314-d4eaf53f-0660-4ccc-a49f-b3adc8d4e00f.png">
<img width="1113" alt="Bildschirmfoto 2022-06-26 um 14 11 00" src="https://user-images.githubusercontent.com/533162/175813511-f5914ed6-5985-41c2-b897-965e503bbe72.png">

**Comment by yarons** [comment]
I see no issues here, thank you.

**Comment by kevinpapst** [comment]
That was quick, thanks 👍 

I am in the middle of a huge refactoring for the next major release. 
Can I sent you a ping once I deployed it to the demo environment for some real-live tests?

**Comment by yarons** [comment]
I'd love that, thank you!

**Comment by kevinpapst** [comment]
@yarons ping 😄 

If you want, you can now test the demo at https://demo-branch.kimai.org/he/login

I created the user `yarons` with the password `password`, but if that is gone after a new installation, you can always use `susan_super` with the password `kitten` and switch the language or create a new super admin account for yourself.

There are missing translations, as the new version ships many changed/new labels, but besides that I hope most elements are now correct working in RTL. 

Happy to receive your feedback!

**Comment by yarons** [comment]
![image](https://user-images.githubusercontent.com/406826/183293421-0dd30231-5924-4d99-b3f5-470ea959ac6d.png)

There are 2 issues here:
1. The arrows are reversed (in My Tasks and Pending Tasks).
2. There's a text overlap in My Projects.

**Reply by kevinpapst** [reply]
Thanks for testing! These points are fixed and demo redeployed.

**Reply by yarons** [reply]
![image](https://user-images.githubusercontent.com/406826/183639729-96554bf5-cd24-4050-90b7-ec6c12e47a09.png)

The names in the tags are unaligned and the format string should be LTR instead of RTL.

**Reply by kevinpapst** [reply]
Thanks again :) I opened a bug for Tomselect (the dropdown component) [here](https://github.com/orchidjs/tom-select/issues/438).

Not sure how to handle the format string though...

**Reply by yarons** [reply]
Just make it LTR regardless of general layout.

Thank you :)

**Reply by kevinpapst** [reply]
Did you try the full way including a generated invoice? 
I tried with some copy&pasted Hebrew character and it looked okay'ish.
But I don't understand the full consequences of mixing RTL and LTR.

**Reply by yarons** [reply]
I tried but it looks like it's not translated at all anyway.

**Reply by kevinpapst** [reply]
Not sure what you are talking about. 
The generated invoice document? 
PDF and HTML should be translated, other formats are only in English (as they are only examples).

**Reply by yarons** [reply]
@kevinpapst Yes, I was talking about the invoice documents, sorry about the delay.

**Reply by kevinpapst** [reply]
I think we should go with bug reports after the release, as this sounds like a minor issue or even misunderstanding.

Some invoice/export document formats (DOCX, Excel) do not support translation, so they are shipped only in english.

**Reply by yarons** [reply]
What does it mean? These are files generated from the software, there must be a way to generate them properly for RTL.

**Reply by kevinpapst** [reply]
They are not really generated, they are loaded and certain replacers are exchanged by the invoice values. 
That means, someone would have to create those template in RTL. I don't have Word and I have no knowledge how to write RTL, so I can't be the one.
For example, see the `company.docx` in this directory: https://github.com/kevinpapst/kimai2/tree/master/templates/invoice/renderer

**Reply by yarons** [reply]
Fair enough, I'll ask my wife's permission to use her copy, is there a recommended version or any will do?

**Reply by kevinpapst** [reply]
The version of Word you mean? I think there are no differences, but I am not certain. This page (of the used library) does not mention anything: https://phpword.readthedocs.io/en/latest/intro.html

**Reply by yarons** [reply]
Just my two cents here, if you don't have access to MS Word, wouldn't it be better to switch to a portable format like ODF for the sake of interoperability? (And also possibly convert it to docx).

**Reply by kevinpapst** [reply]
Who is writing the code to do that switch? I would have already done that, if possible... (there are discussions and issues existing that cover the topic) but the TemplateProcessor which is used, only supports DOCX: https://phpword.readthedocs.io/en/latest/templates-processing.html

**Reply by yarons** [reply]
@kevinpapst Got it, thanks, I'll realign, translate and send it back, do you prefer a PR with the name company-rtl.docx or should I just attach it to this thread?

**Reply by yarons** [reply]
[company-rtl.docx](https://github.com/kevinpapst/kimai2/files/10252709/company-rtl.docx)

This is the file, I'll also submit it as PR, thank you.

**Reply by kevinpapst** [reply]
The best place is https://github.com/kimai/invoice-templates to send a PR to. 
My plan is to include less instead of more templates in the future and then add a link in the UI so people can easily find and download more templates. 

The problem with shipping DOCX templates is, that they can't be translated by Kimai. Users obviously do not know that and then assume that Kimai is buggy. 

To make it simpler, I prepared a directory for translations of that company.docx template (which I will remove from Kimai core).
So you can simple add it to that directory: https://github.com/kimai/invoice-templates/tree/main/docx-company

And please rename it to company-he.docx

**Reply by kevinpapst** [reply]
And btw (for everyone who didn't knew - like me) you can edit DOCX templates for free at https://www.office.com/launch/word

**Reply by kevinpapst** [reply]
I imported the file myself, while I was on it: https://github.com/kimai/invoice-templates/tree/main/docx-company

Thank you for sharing @yarons 👍

**Reply by yarons** [reply]
@kevinpapst Thanks and sorry for the trouble, you're awesome!
